### PR TITLE
[RP Integration] Stop sending requests to all endpoints when ikey is empty

### PR DESCRIPTION
**Pull request by heyams**

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java [hunk 1]
```diff
@@ -87,7 +87,7 @@ private static void mapToConnectionConfiguration(
     // get ikey
     String instrumentationKey = kvps.get(Keywords.INSTRUMENTATION_KEY);
     if (Strings.isNullOrEmpty(instrumentationKey)) {
-      throw new InvalidConnectionStringException("Missing '" + Keywords.INSTRUMENTATION_KEY + "'");
+      logger.warn("New connection string is null or empty");
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:90]: I didn't follow why this change?

**Review by trask — APPROVED** [review]

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java (hunk 1 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:90]: we don't want to throw an exception when instrumentation key is empty.  null or empty is accepted.

Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:90]: maybe in `parseInto` above we can have a separate path empty connection string, that way we don't lose this validation when connection string is not empty, and hopefully we don't need to `warn` in an expected case

separately, i noticed in this method we probably want to reset all the urls in `EndpointProvider` to the defaults before we start populating them, otherwise we could pick up some values left over from the prior connection string

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java [hunk 2]
```diff
@@ -39,9 +39,19 @@
 
   private ConnectionString() {}
 
-  public static void parseInto(String connectionString, TelemetryClient targetConfig)
+  public static void parseInto(String connectionString, TelemetryClient telemetryClient)
       throws InvalidConnectionStringException {
-    mapToConnectionConfiguration(getKeyValuePairs(connectionString), targetConfig);
+    if (Strings.isNullOrEmpty(connectionString)) {
+      telemetryClient.setInstrumentationKey(connectionString);
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:45]: same thing, just since it looks odd to use connection string as an ikey
```suggestion
      telemetryClient.setInstrumentationKey(null);
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java [hunk 3]
```diff
@@ -39,9 +39,19 @@
...
+
+      // reset endpoint to default for ingestion, live metric and profiler
+      EndpointProvider endpointProvider = telemetryClient.getEndpointProvider();
+      endpointProvider.setIngestionEndpoint(endpointProvider.getIngestionEndpoint());
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:49]: this looks like it's just setting the same value back?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/EndpointProvider.java [hunk 4]
```diff
@@ -76,6 +73,14 @@ public URL getAppIdEndpointUrl(String instrumentationKey) {
     return buildAppIdUrl(instrumentationKey);
   }
 
+  // reset endpoint URLs for ingestion, live metric and profiler
+  public void resetEndpointUrls() throws MalformedURLException {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/EndpointProvider.java:77]: 👍

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java (hunk 3 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java]: yup, please check the update. i pushed too quick.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java [hunk 5]
```diff
@@ -39,9 +39,15 @@
...
 
-  public static void parseInto(String connectionString, TelemetryClient targetConfig)
-      throws InvalidConnectionStringException {
-    mapToConnectionConfiguration(getKeyValuePairs(connectionString), targetConfig);
+  public static void parseInto(String connectionString, TelemetryClient telemetryClient)
+      throws InvalidConnectionStringException, MalformedURLException {
+    if (Strings.isNullOrEmpty(connectionString)) {
+      telemetryClient.setInstrumentationKey(null);
+      // reset endpoint to default for ingestion, live metric and profiler
+      telemetryClient.getEndpointProvider().resetEndpointUrls();
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:47]: we should do this also in `mapToConnectionConfiguration` below before starting to overwrite them, since we only conditionally overwrite them below

Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java:47]: in maptoconnectionConfiguration, we should reset all? the above doesn't impact statsbeat and profiler.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/quickpulse/QuickPulsePingSender.java [hunk 6]
```diff
@@ -91,7 +91,7 @@ public QuickPulsePingSender(
 
   public QuickPulseHeaderInfo ping(String redirectedEndpoint) {
     String instrumentationKey = getInstrumentationKey();
-    if (instrumentationKey == null && "java".equals(System.getenv("FUNCTIONS_WORKER_RUNTIME"))) {
+    if (Strings.isNullOrEmpty(instrumentationKey)) {
```
Comment (kryalama) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/quickpulse/QuickPulsePingSender.java:94]: just curios, is there a reason the function runtime check was removed?

Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/quickpulse/QuickPulsePingSender.java:94]: it is no longer applied to function only. azure spring cloud with empty ikey, no live metrics.  that's why we drop that check.

**Review by trask — APPROVED** [review]
